File: net/JetBrains.SignatureVerifier/tests/DefaultRootsTest.cs
using System.Linq;
using JetBrains.SignatureVerifier.Crypt;
using NUnit.Framework;
using Org.BouncyCastle.Asn1.X509;
using Org.BouncyCastle.Math;
using Org.BouncyCastle.Pkix;

namespace JetBrains.SignatureVerifier.Tests
{
  [TestFixture]
  public sealed class DefaultRootsTest
  {
    [Test]
    public void Test()
    {
      (string Name, BigInteger SerialNumber)[] expectedCertificates =
        {
          // @formatter:off
          ("Apple Root CA"                            , new BigInteger("02"                              , 16)),
          ("Certum Trusted Network CA"                , new BigInteger("0444c0"                          , 16)),
          ("Entrust Root Certification Authority - G2", new BigInteger("4a538c28"                        , 16)),
          ("Go Daddy Root Certificate Authority - G2" , new BigInteger("00"                              , 16)),
          ("Microsoft Root Certificate Authority"     , new BigInteger("79ad16a14aa0a5ad4c7358f407132e65", 16)),
          ("Microsoft Root Certificate Authority 2010", new BigInteger("28cc3a25bfba44ac449a9b586b4339aa", 16)),
          ("Microsoft Root Certificate Authority 2011", new BigInteger("3f8bc8b5fc9fb29643b569d66c42e144", 16)),
          ("USERTrust RSA Certification Authority"    , new BigInteger("01fd6d30fca3ca51a81bbc640e35032d", 16))
          // @formatter:on
        };

      (string Name, BigInteger SerialNumber)[] certificates = SignatureVerifier.ResourceUtil.OpenDefaultRoots(codeSignRootsStream =>
        {
          var verificationParams = new SignatureVerificationParams(
            codeSignRootsStream,
            withRevocationCheck: false);
          return verificationParams.RootCertificates.Cast<TrustAnchor>().Select(x => x.TrustedCert).ToList();
        }).Select(x =>
        {
          var name = x.IssuerDN.GetValueList(X509Name.CN).Cast<string>().Single();
          return (name, x.SerialNumber);
        }).OrderBy(x => x.Item1).ToArray();

      Assert.AreEqual(expectedCertificates.Length, certificates.Length);
      for (var n = 0; n < certificates.Length; ++n)
      {
        Assert.AreEqual(expectedCertificates[n].Name, certificates[n].Name);
        Assert.AreEqual(expectedCertificates[n].SerialNumber, certificates[n].SerialNumber);
      }
    }
  }
}

File: jvm/src/test/kotlin/com/jetbrains/signatureverifier/tests/DefaultRootsTest.kt
package com.jetbrains.signatureverifier.tests

import com.jetbrains.signatureverifier.Resources
import com.jetbrains.signatureverifier.crypt.SignatureVerificationParams
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import java.math.BigInteger

class DefaultRootsTest {
  @Test
  fun test() {
    val expectedCertificates = arrayOf(
      // @formatter:off
      Pair("Apple Root CA"                            , BigInteger("02"                              , 16)),
      Pair("Certum Trusted Network CA"                , BigInteger("0444c0"                          , 16)),
      Pair("Entrust Root Certification Authority - G2", BigInteger("4a538c28"                        , 16)),
      Pair("Go Daddy Root Certificate Authority - G2" , BigInteger("00"                              , 16)),
      Pair("Microsoft Root Certificate Authority"     , BigInteger("79ad16a14aa0a5ad4c7358f407132e65", 16)),
      Pair("Microsoft Root Certificate Authority 2010", BigInteger("28cc3a25bfba44ac449a9b586b4339aa", 16)),
      Pair("Microsoft Root Certificate Authority 2011", BigInteger("3f8bc8b5fc9fb29643b569d66c42e144", 16)),
      Pair("USERTrust RSA Certification Authority"    , BigInteger("01fd6d30fca3ca51a81bbc640e35032d", 16))
      // @formatter:on
    )

    val cnRegex = Regex("CN=(?<CN>[^,]*)")
    val certificates = SignatureVerificationParams(
      Resources.GetDefaultRoots(),
      withRevocationCheck = false
    ).RootCertificates!!.map {
      it.trustedCert
    }.map {
      val name = cnRegex.find(it.issuerDN.name)!!.groups["CN"]!!.value
      Pair(name, it.serialNumber)
    }.sortedBy { it.first }.toList()

    Assertions.assertEquals(expectedCertificates.size, certificates.size)
    for (n in certificates.indices) {
      Assertions.assertEquals(expectedCertificates[n].first, certificates[n].first)
      Assertions.assertEquals(expectedCertificates[n].second, certificates[n].second)
    }
  }
}